core/HoistBase.js

```diff
@@ -0,0 +1,275 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+import {XH, Reactive, PersistenceProvider} from '@xh/hoist/core';
+import {throwIf} from '@xh/hoist/utils/js';
+import {observable, extendObservable, runInAction} from '@xh/hoist/mobx';
+import {cloneDeep, isPlainObject, isUndefined} from 'lodash';
+import {PendingTaskModel} from '@xh/hoist/utils/async';
+
+/**
+ * Base class for Services, Models, Stores, and other core entities in Hoist.
+ *
+ * Provides misc support for resource cleanup, state persistence.  Also provides optional
+ * support for lifecycle loading and other useful utilities.
+ *
+ * This class should not typically be extended directly by applications.  Applications should
+ * extend one of its subclasses instead.
+ *
+ * @see HoistModel, HoistService, and Store.
+ */
+export class HoistBase extends Reactive {
+
+    constructor() {
+        super();
+        if (this.implementsLoading) {
+            this._loadModel = new PendingTaskModel();
+            extendObservable(this,
+                {
+                    _lastLoadRequested: null,
+                    _lastLoadCompleted: null,
+                    _lastLoadException: null
+                },
+                {
+                    _lastLoadRequested: observable.ref,
+                    _lastLoadCompleted: observable.ref,
+                    _lastLoadException: observable.ref
+                });
+        }
+    }
+
+    /**
+     * A unique id for this object within the lifetime of this document.
+     * @returns {string}
+     */
+    get xhId() {
+        if (!this._xhId) this._xhId = XH.genId();
+        return this._xhId;
+    }
+
+    /**
+     * Mark an object as managed by this object.
+     *
+     * Managed objects are assumed to hold objects that are created by the referencing object
+     * and therefore should be destroyed when the referencing object is destroyed.
+     *
+     * See also {@see managed}, a decorator that can be used to mark any object held within
+     * a given property as managed.
+     *
+     * @param {object} obj - object to be destroyed
+     * @returns object passed.
+     */
+    markManaged(obj) {
+        this._xhManagedInstances = this._xhManagedInstances ?? [];
+        this._xhManagedInstances.push(obj);
+        return obj;
+    }
+
+    //-----------------------
+    // Load Support
+    //-----------------------
+    /**
+     * Does this object implement loading?
+     * True if an implementation of doLoadAsync been provided.
+     */
+    get implementsLoading() {
```

Comment:
Didn't want to stick with `hasLoadSupport`?  I still like that notion of "support" in that it encompasses a few features that get enabled when `doLoadAsync` has been implemented. (I see we have it still for the section block comment above.)

I will see if I can come up with any alternate suggestions, but I confess I find `implementsLoading` a bit awkward.

Reply:
Will revert to hasLoadSupport?  Agree its the best name, and on sum, the benefits to recycling the name outweigh any confusion.   (had been a bit concerned about developers that were familiar with it as a mixin flag getting confused)

Reply:
Let's discuss the naming -- got a bit more complicated with the latest refactor.